Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+                    .none
+                )
+
+            case .shuttingDown(unclean: let unclean):
+                assert(self.waiters.isEmpty, "Expected to have already cancelled all waiters")
+
+                self.connections.remove(at: index)
+                let isShutdown: StateMachine.ConnectionAction.IsShutdown
+                if self.connections.isEmpty {
+                    self.state = .shutDown
+                    isShutdown = .yes(unclean: unclean)
+                } else {
+                    isShutdown = .no
+                }
+
+                return .init(.none, .closeConnection(connectionState.close(), isShutdown: isShutdown))
+
+            case .shutDown:
+                preconditionFailure("The pool is already shutdown all connections must already been torn down")
+            }
+        }
+
+        /// A connection has been closed
+        mutating func connectionClosed(_ connectionID: Connection.ID) -> Action {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                // because of a race this connection (connection close runs against replace)
+                // was already removed from the state machine.
+                return .init(.none, .none)
+            }
+
+            switch self.state {
+            case .running:
+                guard let nextWaiter = self.waiters.popFirst() else {
+                    self.connections.remove(at: index)
+                    return .init(.none, .none)
+                }
+
+                let closedConnection = self.connections[index]
+                assert(self.connections.count == self.maximumConcurrentConnections,
+                       "Why do we have waiters, if we could open more connections?")
+
+                let eventLoop = nextWaiter.eventLoopRequirement ?? closedConnection.eventLoop
+                let newConnection = HTTP1ConnectionState(
+                    connectionID: self.idGenerator.next(),
+                    eventLoop: eventLoop,
+                    waiter: nextWaiter
+                )
+                self.connections[index] = newConnection
+                return .init(.none, .createConnection(newConnection.connectionID, on: eventLoop))
+
+            case .shuttingDown(unclean: let unclean):
+                assert(self.waiters.isEmpty, "Expected to have already cancelled all waiters")
+
+                self.connections.remove(at: index)
+                if self.connections.isEmpty {
+                    self.state = .shutDown
+                    return .init(.none, .cleanupConnection(close: [], cancel: [], isShutdown: .yes(unclean: unclean)))
+                } else {
+                    return .init(.none, .none)
+                }
+
+            case .shutDown:
+                preconditionFailure("The pool is already shutdown all connections must already been torn down")
+            }
```

Comment:
This entire method seems extremely common to `http1ConnectionReleased`. I think this continues to speak to the question of whether some of this code can be factored out.